### PR TITLE
add the unit test to times.py and fix the bug

### PR DESCRIPTION
**Pull request by Jiasheng-Yang**
Add the unit test to `times.py` and fix the bug; the below is the `pytest` result:

截屏2024-10-22 11 52 48

This is the classic overlap intervals problem of Leedcode, although it can pass the unit test based on provided arguments of the original `__main__` , I still realized that it will generate the invalid time overlaps when (case 1) the short interval is before large interval and (case 2) the short intervals ends after the large interval ends. For example, case 1, the large is 10:00-12:00, the short is 8:30-8:45; case 2, the large is 10:00-12:00, the short is 11:00-12:30. 

I have already modified the original code and pushed the correct one. In addition, I note an easier equation to understand and get the value of `d`.

I learned a lot about the `python.datatime` like `.strftime()` and coverage test. I will continue to review and consolidate my past knowledge about testing, especially in the domain of `mock` and `yaml fixtures for the test`.

Answers UCL-COMP0233-24-25/RSE-Classwork#11

### CODE REVIEW

File: times.py [hunk 1]
```diff
@@ -4,7 +4,10 @@
 def time_range(start_time, end_time, number_of_intervals=1, gap_between_intervals_s=0):
     start_time_s = datetime.datetime.strptime(start_time, "%Y-%m-%d %H:%M:%S")
     end_time_s = datetime.datetime.strptime(end_time, "%Y-%m-%d %H:%M:%S")
+    # I think that it is easier to understand use my below equation:
+    # d = ((end_time_s - start_time_s).total_seconds() - gap_between_intervals_s * (number_of_intervals - 1)) / number_of_intervals
     d = (end_time_s - start_time_s).total_seconds() / number_of_intervals + gap_between_intervals_s * (1 / number_of_intervals - 1)
```
Comment (charlotteesavage) [inline times.py:9]: Just curious: in what way do you think it's easier to understand your equation? I find them pretty similar, but I'm also not familiar with the original leetcode problem

File: times.py [hunk 2]
```diff
@@ -17,10 +20,24 @@ def compute_overlap_time(range1, range2):
...
             low = max(start1, start2)
             high = min(end1, end2)
-            overlap_time.append((low, high))
+            # BUG there is a bug for the original function, it might generate an invalid overlap time, for example
+            # large = time_range("2010-01-12 10:00:00", "2010-01-12 12:00:00")
+            # short = time_range("2010-01-12 8:30:00", "2010-01-12 8:45:00", 2, 60)
+            # Result: [('2010-01-12 10:00:00', '2010-01-12 08:37:00'), ('2010-01-12 10:00:00', '2010-01-12 08:45:00')]
+            # Excepted: []
+            if low < high:
+                overlap_time.append((low, high))
```
Comment (charlotteesavage) [inline times.py:29]: Super interesting find!

File: test_times.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+def test_given_input():
+    large = time_range("2010-01-12 10:00:00", "2010-01-12 12:00:00", number_of_intervals=1, gap_between_intervals_s=0)
+    short = time_range("2010-01-12 10:30:00", "2010-01-12 10:45:00", number_of_intervals=2, gap_between_intervals_s=60)
+
+    result = compute_overlap_time(large, short)
+    expected = [('2010-01-12 10:30:00', '2010-01-12 10:37:00'), ('2010-01-12 10:38:00', '2010-01-12 10:45:00')]
+
+    # judge that the result matches the expected value
+    assert result == expected, f"Expected: {expected}, but Result: {result}, doesn't match!"
```
Comment (charlotteesavage) [inline test_times.py:12]: In what way do they not match?

File: times.py (hunk 1 above)
Comment (Jiasheng-Yang) [reply times.py:9]: It's very similar to this [problem](https://leetcode.com/problems/merge-intervals/description/), except this exercise adds some manipulation of time. The equation is `d = ((end_time_s - start_time_s).total_seconds() - gap_between_intervals_s * (number_of_intervals - 1)) / number_of_intervals`

**Pull request closed without merging**